Assets/MixedRealityToolkit.SDK/Experimental/Features/Support/Scripts/BaseServiceManager.cs

```diff
@@ -0,0 +1,395 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Utilities;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit
+{
+    /// <summary>
+    /// Base class providing service registration and management functionality. This class can be used to implement a
+    /// custom service management component for one or more services, similar to the MixedRealityToolkit object.
+    /// </summary>
+    public class BaseServiceManager : MonoBehaviour, IMixedRealityServiceRegistrar
+    {
+        /// <summary>
+        /// The collection of registered services.
+        /// </summary>
+        protected Dictionary<Type, IMixedRealityService> registeredServices = new Dictionary<Type, IMixedRealityService>();
+
+        /// <summary>
+        /// The collection of registered data providers.
+        /// </summary>
+        private List<IMixedRealityDataProvider> dataProviders = new List<IMixedRealityDataProvider>();
+
+        #region MonoBehaviour implementation
+
+        protected virtual void Update()
+        {
+            if (Application.isPlaying)
+            {
+                IReadOnlyList<IMixedRealityService> serviceSnapshot = new List<IMixedRealityService>(registeredServices.Values);
+                for (int i = 0; i < serviceSnapshot.Count; i++)
+                {
+                    serviceSnapshot[i]?.Update();
```

Comment:
> serviceSnapshot[i]?.Update(); [](start = 19, length = 30)

Is there a good reason why we would continue to store null values in registeredServices?

It doesn't make that much sense to me that we do null checks on each thing - we should just not store null values as they aren't useful.

Reply:
was a safely check. there shouldnt ever be nulls

Reply:
fixed